File: Lab/Lab_5/Java/Lab_5_1.java
import java.util.Scanner;

public class Lab_5_1{

    public static String longestNiceSubstring(String str) {
        
        if(str.isEmpty())
            return "";
        
        boolean lower[] = new boolean[26], upper[] = new boolean[26];

        for(char i : str.toCharArray()){
            if(i < 'a')
                upper[i - 'A'] = true;
            else
                lower[i - 'a'] = true;
        }

        for(int i=0; i<str.length(); i++)
        {
            int boolIndex = str.charAt(i) < 'a' ? str.charAt(i) - 'A' : str.charAt(i) - 'a';

            if(lower[boolIndex] != upper[boolIndex])
            {
                String left = longestNiceSubstring(str.substring(0,i));
                String right = longestNiceSubstring(str.substring(i + 1));

                return left.length() < right.length() ? right : left;
            }
        }
        return str;
    }

    public static void main(String[] args)
    {
        Scanner scan = new Scanner(System.in);
        
        String str = scan.nextLine();
        System.out.println(longestNiceSubstring(str));
    }
}

File: Lab/Lab_5/Cpp/Lab_5_1.cpp
#include <bits/stdc++.h>

using namespace std;

class Lab_5_1 {
    public:
        string longestNiceSubstring(string str)
        {
            if(str.length() < 2)
                return "";

            bool upper[26] = {false}, lower[26] = {false};

            for(char i: str)
                if(i < 97)
                    upper[i - 'A'] = true;
                else
                    lower[i - 'a'] = true;

            for(int i=0; i<str.length(); i++){
                int boolindex = str[i] < 97 ? str[i] - 'A': str[i] - 'a';

                if(upper[boolindex] != lower[boolindex]){
                    string left = longestNiceSubstring(str.substr(0, i));
                    string right = longestNiceSubstring(str.substr(i + 1));

                    return left.length() < right.length() ? right: left;
                }
            }
            return str;
        }
};

int main(){
    string str;
    cin >> str;

    Lab_5_1 sol ;
    cout<< sol.longestNiceSubstring(str);
}
                